core/kernel/generic_pm.c

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#include <keep.h>
+#include <kernel/generic_pm.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <types_ext.h>
+
+/*
+ * Generic power management support for platforms and drivers.
+ * Core components can register callback for suspend and resume
+ * sequences. Callback can be related to a specific subsystem,
+ * based on an opaque identifier. Callback has some ordering
+ * directives as specified by the PM_CB_LEVEL_xxx value.
+ *
+ * Platform specific power sequence can call pm_suspend_resume()
+ * for a suspend or resume operation and for a target subsystem.
+ * Registered callbacks for that subsystem are called with the
+ * ordering directive.
+ */
+
+static bool __maybe_unused cb_is_valid(int (*cb)(enum pm_op op, void *hdl),
+				       void *hdl)
+{
+	paddr_t cb_phy = virt_to_phys(cb);
+	paddr_t hdl_phy = virt_to_phys(hdl);
+	bool __maybe_unused valid;
+
+	valid = (phys_to_virt(cb_phy, MEM_AREA_TEE_RAM_RX) == cb) &&
+		(!hdl_phy ||
+		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RX) == hdl) ||
+		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RO) == hdl) ||
+		 (phys_to_virt(hdl_phy, MEM_AREA_TEE_RAM_RW) == hdl));
+
+	if (!valid) {
+		EMSG("pm_cb mandates unpaged arguments %p %p", (void *)cb, hdl);
+	}
+
+	return valid;
+}
+
+static bool __maybe_unused cb_level_is_valid(enum pm_op op)
+{
+	switch (op) {
+	case PM_OP_SUSPEND:
+	case PM_OP_RESUME:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool __maybe_unused pm_subsys_is_valid(unsigned int pm_subsys)
+{
+	return pm_subsys <= UINT16_MAX;
+}
+
+struct pm_cb {
+	int (*callback)(enum pm_op op, void *handle);
+	void *handle;
+	uint8_t level;
+	uint8_t suspended;
+	uint16_t subsys;
+};
+
+static struct pm_cb *pm_cb_ref;
+static size_t pm_cb_count;
+
+void register_pm_cb2(int (*callback)(enum pm_op op, void *handle),
+		     void *handle,
+		     enum pm_cb_level cb_level,
+		     unsigned int pm_subsys)
+{
+	assert(cb_is_valid(callback, handle) &&
+	       cb_level_is_valid(cb_level) &&
+	       pm_subsys_is_valid(pm_subsys));
+
+	size_t count = pm_cb_count;
+	struct pm_cb *pm_cb = pm_cb_ref;
+
+	pm_cb = realloc(pm_cb, sizeof(struct pm_cb) * (count + 1));
+	if (!pm_cb)
+		panic();
+
+	memset(&pm_cb[count], 0, sizeof(struct pm_cb));
```

Comment:
How about writing this as:
```
pm_cb[count] = (struct pm_cb){.callback = callback, .handle = handle,
                              .level = cb_level, .subsys = pm_subsys, };
```
We haven't used this pattern much, but it would provide better type safety and is a bit more compact while quite readable. Also note that I skipped the casts as they aren't needed.

Reply:
right.